File: km003c-slint/java/dev/okhsunrog/km003c/DynamicColors.java
package dev.okhsunrog.km003c;

import android.content.Context;
import android.os.Build;

/**
 * The Material 3 colour roles of the system's dynamic theme.
 *
 * <p>Android 14 exposes every role of the wallpaper-derived scheme as a colour
 * resource, in a light and a dark variant. The names here are the SDK's own
 * symbols, so {@code javac} checks them; transcribed into Rust they would be
 * strings nothing checks until they run.
 *
 * <p>The whole scheme crosses to Rust as one string, {@code variant.role=AARRGGBB}
 * per line, which costs one JNI call instead of one per colour. Role names are
 * those of Slint's {@code MaterialScheme}. Roles the system does not provide
 * (the inverse roles, shadow, scrim) are left out and keep the library's
 * defaults.
 */
public final class DynamicColors {

    private final Context context;
    private final StringBuilder out = new StringBuilder(4096);

    public DynamicColors(Context context) {
        this.context = context;
    }

    /** The scheme, or an empty string before Android 14. */
    public String scheme() {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.UPSIDE_DOWN_CAKE) {
            return "";
        }
        out.setLength(0);

        role("primary", android.R.color.system_primary_light, android.R.color.system_primary_dark);
        // Material defines surface tint as the primary colour.
        role("surfaceTint", android.R.color.system_primary_light, android.R.color.system_primary_dark);
        role("onPrimary", android.R.color.system_on_primary_light, android.R.color.system_on_primary_dark);
        role("primaryContainer", android.R.color.system_primary_container_light,
                android.R.color.system_primary_container_dark);
        role("onPrimaryContainer", android.R.color.system_on_primary_container_light,
                android.R.color.system_on_primary_container_dark);

        role("secondary", android.R.color.system_secondary_light, android.R.color.system_secondary_dark);
        role("onSecondary", android.R.color.system_on_secondary_light, android.R.color.system_on_secondary_dark);
        role("secondaryContainer", android.R.color.system_secondary_container_light,
                android.R.color.system_secondary_container_dark);
        role("onSecondaryContainer", android.R.color.system_on_secondary_container_light,
                android.R.color.system_on_secondary_container_dark);

        role("tertiary", android.R.color.system_tertiary_light, android.R.color.system_tertiary_dark);
        role("onTertiary", android.R.color.system_on_tertiary_light, android.R.color.system_on_tertiary_dark);
        role("tertiaryContainer", android.R.color.system_tertiary_container_light,
                android.R.color.system_tertiary_container_dark);
        role("onTertiaryContainer", android.R.color.system_on_tertiary_container_light,
                android.R.color.system_on_tertiary_container_dark);

        role("error", android.R.color.system_error_light, android.R.color.system_error_dark);
        role("onError", android.R.color.system_on_error_light, android.R.color.system_on_error_dark);
        role("errorContainer", android.R.color.system_error_container_light,
                android.R.color.system_error_container_dark);
        role("onErrorContainer", android.R.color.system_on_error_container_light,
                android.R.color.system_on_error_container_dark);

        role("background", android.R.color.system_background_light, android.R.color.system_background_dark);
        role("onBackground", android.R.color.system_on_background_light, android.R.color.system_on_background_dark);
        role("surface", android.R.color.system_surface_light, android.R.color.system_surface_dark);
        role("onSurface", android.R.color.system_on_surface_light, android.R.color.system_on_surface_dark);
        role("surfaceVariant", android.R.color.system_surface_variant_light,
                android.R.color.system_surface_variant_dark);
        role("onSurfaceVariant", android.R.color.system_on_surface_variant_light,
                android.R.color.system_on_surface_variant_dark);
        role("outline", android.R.color.system_outline_light, android.R.color.system_outline_dark);
        role("outlineVariant", android.R.color.system_outline_variant_light,
                android.R.color.system_outline_variant_dark);

        role("surfaceDim", android.R.color.system_surface_dim_light, android.R.color.system_surface_dim_dark);
        role("surfaceBright", android.R.color.system_surface_bright_light,
                android.R.color.system_surface_bright_dark);
        role("surfaceContainerLowest", android.R.color.system_surface_container_lowest_light,
                android.R.color.system_surface_container_lowest_dark);
        role("surfaceContainerLow", android.R.color.system_surface_container_low_light,
                android.R.color.system_surface_container_low_dark);
        role("surfaceContainer", android.R.color.system_surface_container_light,
                android.R.color.system_surface_container_dark);
        role("surfaceContainerHigh", android.R.color.system_surface_container_high_light,
                android.R.color.system_surface_container_high_dark);
        role("surfaceContainerHighest", android.R.color.system_surface_container_highest_light,
                android.R.color.system_surface_container_highest_dark);

        // Fixed roles are the same in both variants.
        fixed("primaryFixed", android.R.color.system_primary_fixed);
        fixed("primaryFixedDim", android.R.color.system_primary_fixed_dim);
        fixed("onPrimaryFixed", android.R.color.system_on_primary_fixed);
        fixed("onPrimaryFixedVariant", android.R.color.system_on_primary_fixed_variant);
        fixed("secondaryFixed", android.R.color.system_secondary_fixed);
        fixed("secondaryFixedDim", android.R.color.system_secondary_fixed_dim);
        fixed("onSecondaryFixed", android.R.color.system_on_secondary_fixed);
        fixed("onSecondaryFixedVariant", android.R.color.system_on_secondary_fixed_variant);
        fixed("tertiaryFixed", android.R.color.system_tertiary_fixed);
        fixed("tertiaryFixedDim", android.R.color.system_tertiary_fixed_dim);
        fixed("onTertiaryFixed", android.R.color.system_on_tertiary_fixed);
        fixed("onTertiaryFixedVariant", android.R.color.system_on_tertiary_fixed_variant);

        return out.toString();
    }

    private void role(String name, int light, int dark) {
        entry("light", name, light);
        entry("dark", name, dark);
    }

    private void fixed(String name, int id) {
        role(name, id, id);
    }

    private void entry(String variant, String name, int id) {
        out.append(variant).append('.').append(name).append('=')
                .append(Integer.toHexString(context.getColor(id))).append('\n');
    }
}

File: km003c-slint/src/dynamic_colors.rs
//! The system's dynamic Material colours, on Android 14 and later.
//!
//! `java/.../DynamicColors.java` reads every colour role of the wallpaper-derived
//! scheme and hands them over as one `variant.role=AARRGGBB` string. The roles
//! it provides replace those of the Material library's default palette; the
//! rest keep their defaults. The library then picks the light or dark scheme
//! from the system theme itself.

use std::collections::HashMap;

use slint::Color;

use crate::MaterialScheme;

/// Overwrite the roles of `scheme` that `listing` gives for `variant`
/// (`light` or `dark`). Returns how many roles were set.
#[cfg_attr(not(target_os = "android"), allow(dead_code))]
pub fn apply_roles(scheme: &mut MaterialScheme, listing: &str, variant: &str) -> usize {
    let roles: HashMap<&str, Color> = listing
        .lines()
        .filter_map(|line| {
            let (key, hex) = line.split_once('=')?;
            let (line_variant, role) = key.split_once('.')?;
            let argb = u32::from_str_radix(hex, 16).ok()?;
            (line_variant == variant).then_some((role, Color::from_argb_encoded(argb)))
        })
        .collect();

    let mut applied = 0;
    macro_rules! set_roles {
        ($($field:ident),* $(,)?) => {
            $(
                if let Some(color) = roles.get(stringify!($field)) {
                    scheme.$field = *color;
                    applied += 1;
                }
            )*
        };
    }
    set_roles!(
        primary,
        surfaceTint,
        onPrimary,
        primaryContainer,
        onPrimaryContainer,
        secondary,
        onSecondary,
        secondaryContainer,
        onSecondaryContainer,
        tertiary,
        onTertiary,
        tertiaryContainer,
        onTertiaryContainer,
        error,
        onError,
        errorContainer,
        onErrorContainer,
        background,
        onBackground,
        surface,
        onSurface,
        surfaceVariant,
        onSurfaceVariant,
        outline,
        outlineVariant,
        surfaceDim,
        surfaceBright,
        surfaceContainerLowest,
        surfaceContainerLow,
        surfaceContainer,
        surfaceContainerHigh,
        surfaceContainerHighest,
        primaryFixed,
        primaryFixedDim,
        onPrimaryFixed,
        onPrimaryFixedVariant,
        secondaryFixed,
        secondaryFixedDim,
        onSecondaryFixed,
        onSecondaryFixedVariant,
        tertiaryFixed,
        tertiaryFixedDim,
        onTertiaryFixed,
        onTertiaryFixedVariant,
    );
    applied
}

#[cfg(target_os = "android")]
pub use android::apply;

#[cfg(target_os = "android")]
mod android {
    use jni::objects::{JObject, LoaderContext};
    use jni::{JavaVM, bind_java_type};
    use slint::ComponentHandle;
    use slint::android::AndroidApp;
    use tracing::{info, warn};

    use super::apply_roles;
    use crate::{App, MaterialPalette};

    bind_java_type! {
        Context => android.content.Context,
        methods {
            fn get_class_loader {
                name = "getClassLoader",
                sig = () -> JClassLoader,
            },
        },
    }

    bind_java_type! {
        DynamicColors => "dev.okhsunrog.km003c.DynamicColors",
        type_map = {
            Context => "android.content.Context",
        },
        constructors {
            fn new(context: Context),
        },
        methods {
            fn scheme {
                name = "scheme",
                sig = () -> JString,
            },
        },
    }

    /// Install the system's dynamic colours, keeping the default palette when
    /// they are unavailable.
    pub fn apply(ui: &App, app: &AndroidApp) {
        let listing = match read(app) {
            Ok(listing) => listing,
            Err(error) => {
                warn!("Could not read the system's dynamic colours: {error}");
                return;
            }
        };
        if listing.is_empty() {
            info!("Dynamic colours need Android 14; keeping the default palette");
            return;
        }

        let palette = ui.global::<MaterialPalette>();
        let mut schemes = palette.get_schemes();
        let applied =
            apply_roles(&mut schemes.light, &listing, "light") + apply_roles(&mut schemes.dark, &listing, "dark");
        palette.set_schemes(schemes);
        info!("Applied {applied} dynamic colour roles");
    }

    fn read(app: &AndroidApp) -> Result<String, jni::errors::Error> {
        // Slint has usually created the JavaVM singleton by now; adopt the
        // process VM if it has not.
        if JavaVM::singleton().is_err() {
            // SAFETY: android-activity documents vm_as_ptr() as the process
            // JavaVM, which is what JavaVM::from_raw expects.
            unsafe { JavaVM::from_raw(app.vm_as_ptr() as *mut _) };
        }

        JavaVM::singleton()?.attach_current_thread(|env| {
            // SAFETY: activity_as_ptr() is the Activity jobject, alive for as
            // long as the activity is.
            let activity = unsafe { JObject::from_raw(env, app.activity_as_ptr() as *mut _) };
            let context = Context::cast_local(env, activity)?;

            // A thread attached from native code finds classes through the
            // system class loader, which does not know the app's own classes.
            let loader = context.get_class_loader(env)?;
            DynamicColorsAPI::get(env, &LoaderContext::Loader(&loader))?;

            let colors = DynamicColors::new(env, &context)?;
            let listing = colors.scheme(env)?;
            listing.try_to_string(env)
        })
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn roles_of_the_requested_variant_replace_the_defaults() {
        let mut scheme = MaterialScheme::default();
        let listing = "light.primary=ff112233\ndark.primary=ff445566\nlight.onSurface=80aabbcc\n";

        let applied = apply_roles(&mut scheme, listing, "light");

        assert_eq!(applied, 2);
        assert_eq!(scheme.primary, Color::from_argb_u8(0xff, 0x11, 0x22, 0x33));
        assert_eq!(scheme.onSurface, Color::from_argb_u8(0x80, 0xaa, 0xbb, 0xcc));
    }

    #[test]
    fn unknown_roles_and_malformed_lines_are_ignored() {
        let mut scheme = MaterialScheme::default();
        let listing = "light.notARole=ff000000\nlight.primary=nothex\ngarbage\n";

        assert_eq!(apply_roles(&mut scheme, listing, "light"), 0);
        assert_eq!(scheme, MaterialScheme::default());
    }
}
